### PR TITLE
🚦 Command filters!

### PR DESCRIPTION
**Pull request by binaryoverload**
# Pull request

 - [x] I have read and agreed to the [code of conduct](https://github.com/CascadeBot/CascadeBot/blob/master/.github/CODE_OF_CONDUCT.md).
 - [x] I have read and complied with the [contributing guidelines](https://github.com/CascadeBot/CascadeBot/blob/master/.github/CONTRIBUTING.md).
 - [x] What I'm implementing was assigned to me or was not being worked on and is a somewhat requested feature. For reference see our [GitHub projects](https://github.com/orgs/CascadeBot/projects).
 - [x] I have tested all of my changes.
 
## Added/Changed
 - Command filters 
    - Filter a command by the user, role or channel
    - Both blacklist and whitelist is supported
    - For filtering, users can select that all conditions have to be correct (AND) or only one of them (OR)

## TODO
- [x] `;filters commands`
- [x] `;filters list`
- [x] `;filters roles`
- [x] `;filters users`
- [x] Proper message / not message for when commands are blocked

### CODE REVIEW

File: src/main/java/org/cascadebot/cascadebot/events/CommandListener.java [hunk 1]
```diff
@@ -127,6 +128,14 @@ private void processCommands(GuildMessageReceivedEvent event, GuildData guildDat
             if (!isAuthorised(cmd, context)) {
                 return;
             }
+
+            if (!processFilters(cmd, context)) {
+                // TODO: Moderation log event
+                // TODO: Message for the end user?
+                context.getTypedMessaging().replyDanger("Message blocked");
```
Comment (JamesHawkinss) [inline src/main/java/org/cascadebot/cascadebot/events/CommandListener.java:135]: Should probably languify this

File: src/main/kotlin/org/cascadebot/cascadebot/commands/management/filters/FiltersCommandsSubCommand.kt [hunk 2]
```diff
@@ -0,0 +1,9 @@
+/*
+ * Copyright (c) 2020 CascadeBot. All rights reserved.
+ * Licensed under the MIT license.
+ */
+
+package org.cascadebot.cascadebot.commands.management.filters
+
+class FiltersCommandsSubCommand {
```
Comment (JamesHawkinss) [inline src/main/kotlin/org/cascadebot/cascadebot/commands/management/filters/FiltersCommandsSubCommand.kt:8]: what a lonely class :(

File: src/main/kotlin/org/cascadebot/cascadebot/commands/management/filters/FiltersListSubCommand.kt [hunk 3]
```diff
@@ -0,0 +1,9 @@
+/*
+ * Copyright (c) 2020 CascadeBot. All rights reserved.
+ * Licensed under the MIT license.
+ */
+
+package org.cascadebot.cascadebot.commands.management.filters
+
+class FiltersListSubCommand {
```
Comment (JamesHawkinss) [inline src/main/kotlin/org/cascadebot/cascadebot/commands/management/filters/FiltersListSubCommand.kt:8]: another lonely class :((

File: src/main/kotlin/org/cascadebot/cascadebot/data/objects/CommandFilter.kt [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+            FilterType.WHITELIST -> if (combinedResult) FilterResult.ALLOW else FilterResult.DENY
+            FilterType.BLACKLIST -> if (combinedResult) FilterResult.DENY else FilterResult.ALLOW
+        }
+    }
+
+    val filterEmbed: EmbedBuilder
+        get() = EmbedBuilder()
+
+    /**
+     * Determines whether users who match this filter will be blocked or whitelisted
```
Comment (JamesHawkinss) [inline src/main/kotlin/org/cascadebot/cascadebot/data/objects/CommandFilter.kt:72]: this is just me being stupid, but maybe make a separate file for each of these enums? probably ott

Comment (binaryoverload) [reply src/main/kotlin/org/cascadebot/cascadebot/data/objects/CommandFilter.kt:72]: The answer there is.... not sure?

File: src/main/kotlin/org/cascadebot/cascadebot/commands/management/filters/FiltersListSubCommand.kt (hunk 3 above)
Comment (binaryoverload) [reply src/main/kotlin/org/cascadebot/cascadebot/commands/management/filters/FiltersListSubCommand.kt:8]: TODO

File: src/main/kotlin/org/cascadebot/cascadebot/commands/management/filters/FiltersCommandsSubCommand.kt (hunk 2 above)
Comment (binaryoverload) [reply src/main/kotlin/org/cascadebot/cascadebot/commands/management/filters/FiltersCommandsSubCommand.kt:8]: TODO :(

File: src/main/java/org/cascadebot/cascadebot/events/CommandListener.java (hunk 1 above)
Comment (binaryoverload) [reply src/main/java/org/cascadebot/cascadebot/events/CommandListener.java:135]: Yes I agree

**Comment by binaryoverload** [comment]
@weeryan17 Ready to review, just need to finish a language string 👌

File: src/main/java/org/cascadebot/cascadebot/data/language/Locale.java [hunk 5]
```diff
@@ -46,6 +46,10 @@ public InputStream getLanguageFile() {
         return CascadeBot.class.getClassLoader().getResourceAsStream("lang/" + getLanguageFileName());
     }
 
+    public String i18n(String path, Object... objects) {
+        return Language.i18n(this, path, objects);
+    }
+
```
Comment (weeryan17) [inline src/main/java/org/cascadebot/cascadebot/data/language/Locale.java:52]: WOOOOOO!

File: src/main/kotlin/org/cascadebot/cascadebot/data/objects/GuildSettingsManagement.kt [hunk 6]
```diff
@@ -9,7 +9,14 @@ class GuildSettingsManagement {
     @Setting
     var allowTagCommands = true
 
+    @Setting
+    var displayFilterError = false
```
Comment (weeryan17) [inline src/main/kotlin/org/cascadebot/cascadebot/data/objects/GuildSettingsManagement.kt:13]: so you do have a setting, but you don't use it.....

Comment (weeryan17) [reply src/main/kotlin/org/cascadebot/cascadebot/data/objects/GuildSettingsManagement.kt:13]: Unless I'm mistaken what this setting is for.

File: src/main/java/org/cascadebot/cascadebot/events/CommandListener.java [hunk 7]
```diff
@@ -127,6 +128,17 @@ private void processCommands(GuildMessageReceivedEvent event, GuildData guildDat
             if (!isAuthorised(cmd, context)) {
                 return;
             }
+
+            if (!processFilters(cmd, context)) {
+                // TODO: Moderation log event
+                if (context.getData().getManagement().getDisplayFilterError()) {
+                    var message = context.i18n("commands.filters.message_blocked") +
+                            (context.getMember().hasPermission(Permission.ADMINISTRATOR) ? context.i18n("commands.filters.message_blocked_admin") : "");
+                    context.getTypedMessaging().replyDanger(message);
```
[the remark this answers is not in the document]
Comment (binaryoverload) [reply src/main/java/org/cascadebot/cascadebot/events/CommandListener.java:137]: Agree about auto-deletion but there is already a setting? See line 134

[the remark this answers is not in the document]
Comment (weeryan17) [reply src/main/java/org/cascadebot/cascadebot/events/CommandListener.java:137]: Whoops I missed that.

**Comment by binaryoverload** [comment]
;jenkins build

**Comment by binaryoverload** [comment]
jenkins build

**Pull request merged**